---
title: "Using regulatory genomics data to interpret the function of disease variants and prioritise genes from expression studies"
author:
  name: Enrico Ferrero
  email: enrico.x.ferrero@gsk.com
  affiliation: Computational Biology, GSK, Medicines Research Centre, Gunnels Wood Road, Stevenage, SG1 2NY, UK
biblio-style: unsrtnat
natbiboptions: super
csl: biomed-central.csl
bibliography: references.bib
output:
  BiocWorkflowTools::f1000_article: default
  github_document:
    html_preview: no
  html_document: default
  word_document: default
always_allow_html: yes
editor_options:
  chunk_output_type: console
keywords: bioconductor; r; rstats; regulatory genomics; functional genomics; genetics; gwas; transcriptomics; integration; multiomics
abstract: The identification of therapeutic targets is a critical step in the research and developement of new drugs, with several drug discovery programmes failing because of a weak linkage between target and disease. Genome-wide association studies and large-scale gene expression experiments are providing insights into the biology of several common and complex diseases, but the complexity of transcriptional regulation mechanisms often limit our understanding of how genetic variation can influence changes in gene expression. Several initiatives in the field of regulatory genomics are aiming to close this gap by systematically identifying and cataloguing regulatory elements such as promoters and enhacers across different tissues and cell types. In this Bioconductor workflow, we will explore how different types of regulatory genomic data can be used for the functional interpretation of disease-associated variants and for the prioritisation of gene lists from gene expression experiments.
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

# Introduction

Discovering and bringing new drugs to the market is a long, expensive and inefficient process [@Waring2015; @DiMasi2016].
The majority of drug discovery programmes fail for efficacy reasons [@Harrison2016], with up to 40% of these failures due to lack of a clear link between the target and the disease under investigation [@Cook2014].

Target selection, the first step in drug discovery programmes, is thus a critical decision point.
It has previously been shown that therapeutic targets with a genetic link to the disease under investigation are more likely to progress through the drug discovery pipeline, suggesting that genetics can be used as a tool to prioritise and validate drug targets in early discovery [@Plenge2013; @Nelson2015].

One of the biggest challenges in translating findings from genome-wide association studies (GWASs) to therapies is that the great majority of single nucleotide polymorphisms (SNPs) associated with disease are found in non-coding regions of the genome, and therefore cannot be easily linked to a target gene [@Maurano2012].
Many of these SNPs could be regulatory variants, affecting the expression of nearby or distal genes by interfering with the transcriptional process [@Ward2012].

The most established way to map disease-associated regulatory variants to target genes is to use expression quantitative trait loci (eQTLs) [@Albert2015], variants that affect the expression of specific genes.
The GTEx consortium profiled eQTLs across 44 human tissues by performing a large-scale mapping of genome-wide correlations between genetic variants and gene expression [@GTEx2017].

However, depending on the power of the study, it might not be possible to detect all existing regulatory variants as eQTLs.
An alternative is to use information on the location of promoters and distal enhancers across the genome and link these regulatory elements to their target genes.
Large, multi-centre initiatives such as ENCODE [@ENCODE2012], Roadmap Epigenomics [@Roadmap2015] and BLUEPRINT [@Adams2012; @Stunnenberg2016] mapped regulatory elements in the genome by profiling a number of chromatin features, including DNase hypersensitive sites (DHSs), several types of histone marks and binding of chromatin-associated proteins in a large number of cells and tissues.
Similarly, the FANTOM consortium used cap analysis of gene expression (CAGE) to identify promoters and enhancers across hundreds of cells and tissues [@Fantom2014].

Knowing that a certain stretch of DNA is an enhancer is however not informative of the target gene(s).
One way to infer links between enhancers and promoters *in silico* is to identify significant correlations across a large panel of cell types, an approach that was used for distal and promoter DHSs [@Thurman2012] as well as for CAGE-defined promoters and enhancers [@Andersson2014].
Experimental methods to assay interactions between regulatory elements also exist.
Chromatin interaction analysis by paired-end tag sequencing (ChIA-PET) [@Fullwood2009; @Zhang2013] couples chromatin immunoprecipitation with DNA ligation to identify DNA regions interacting thanks to the binding of a specific protein.
Promoter capture Hi-C [@Mifsud2015; @Javierre2016] extends chromatin conformation capture by using "baits" to enrich for promoter interactions and increase resolution.

Overall, linking genetic variants to their candidate target genes is not straightforward, not only because of the complexity of the human genome and transcriptional regulation, but also because of the variety of data types and approaches that can be used.
To address this problem, we developed STOPGAP, a database of disease variants mapped to their most likely target gene(s) using several different types of regulatory genomic data [@Shen2017].
The database is currently undergoing a major overhaul and will eventually be superseded by [POSTGAP](https://github.com/Ensembl/postgap).
A valid and recent alternative is INFERNO [@Amlie-Wolf2017], though it does only rely on eQTL data for target gene assignment.
These resources implement some or all of the approaches that will be reviewed in the workflow and constitute good entry points for identifying the most likely target gene(s) of regulatory SNPs. However, as they tend to hide much of the complexity involved in the process, we will not use them and rely on the original datasets instead.

In this workflow, we will explore how regulatory genomic data can be used to connect the genetic and transcriptional layers by providing a framework for the discovery of novel therapeutic targets.
We will use eQTL data from GTEx [@GTEx2017], FANTOM5 correlations between promoters and enhancers [@Andersson2014] and promoter capture Hi-C data [@Javierre2016] to annotate significant GWAS variants to putative target genes and to prioritise genes obtained from a differential expression analysis (Figure 1).

```{r figure1, fig.cap = 'Diagram showing a schematic representation of the workflow and the steps involved.', echo = FALSE}
# uncomment the following line to install package
# install.packages("DiagrammeR")
library(DiagrammeR)
mermaid("
  graph TB

    A[<strong>RNA-seq data</strong>] --> B[<strong>DEGs in SLE vs healthy blood</strong>]
    C[<strong>GWAS data</strong>] --> D[<strong>Significant SLE GWAS SNPs</strong>]
    B --> E
    D --> E[<strong><center>Annotate coding, promoter<br>and UTR SNPs to target genes</center></strong>]
    E --> F[<strong><center>Use GTEx blood eQTL data<br>to annotate intronic and <br> intergenic SNPs to target genes</center></strong>]
    F --> G[<strong><center>Use FANTOM5 correlations between <br> CAGE-defined enhancers and promoters <br> to annotate intronic and intergenic <br> SNPs to target genes</center></strong>]
    G --> H[<strong><center>Use promoter capture Hi-C data <br> from primary hematopoietic cells <br> to annotate intronic and <br> intergenic SNPs to target genes</center></strong>]
    H --> I[<strong><center>Prioritised hits <br> for target discovery</center></strong>]

classDef default fill:#FFFFFF,stroke:#F2673C,stroke-width:4px
")
```

# Workflow

## Install required packages

R version 3.4.2 and Bioconductor version 3.6 were used for the analysis.
The code below will install all required packages and dependencies from Bioconductor and CRAN:

```{r}
source("https://bioconductor.org/biocLite.R")
# uncomment the following line to install packages
#biocLite(c("clusterProfiler", "DESeq2", "GenomicFeatures", "GenomicInteractions", "GenomicRanges", "ggplot2", "Gviz", "gwascat", "InteractionSet", "recount", "pheatmap", "RColorBrewer", "rtracklayer", "R.utils", "splitstackshape", "VariantAnnotation"))
```

## Gene expression data and differential gene expression analysis

We start with a common scenario: we ran a RNA-seq experiment comparing patients with a disease and healthy individuals, and would like to discover key disease genes and potential therapeutic targets by integrating genetic information in our analysis.

The RNA-seq data we will be using comes from blood of patients with systemic lupus erythematosus (SLE) and healthy controls [@Hung2015].
SLE is a chronic autoimmune disorder that can affect several organs with a significant unmet medical need [@Kaul2016].
It is a complex and remarkably heterogeneous disease, in terms of both genetics and clinical manifestations [@Marion2014].
Early diagnosis and classification of SLE remain extremely challenging [@Amezcua-Guerra2015].

In the original study [@Hung2015], the authors explore transcripts bound by Ro60, an RNA-binding protein against which some SLE patients produce autoantibodies.
They identify Alu retroelements among these transcripts and use RNA-seq data to check their expression levels, observing that Alu elements are significantly more expressed in SLE patients, and particularly in those patients with anti-Ro antibodies and with a higher interferon signature metric (ISM). 

We are going to use `recount` [@Collado-Torres2017] to obtain gene-level counts:

```{r}
library(recount)
# uncomment the following line to download dataset
#download_study("SRP062966")
load(file.path("SRP062966", "rse_gene.Rdata"))
rse <- scale_counts(rse_gene)
```

Other Bioconductor packages that can be used to access data from gene expression experiments directly in R are `GEOquery` [@Davis2007] and `ArrayExpress` [@Kauffmann2009]. 

We have `r ncol(rse)` samples overall.
This is what the matrix of counts looks like:

```{r}
assay(rse)[1:3, 1:3]
```

Each gene is a row and each sample is a column.
We note that genes are annotated using the GENCODE [@Harrow2012] v25 annotation, which will be useful later on.

To check how we can split samples between cases and controls, we can have a look at the metadata contained in the `characteristics` column, which is a `CharacterList` object:

```{r}
head(rse$characteristics, 3)
```

We have information about the disease status of the sample, the tissue of origin, the presence and level of anti-ro autoantibodies and the value of the ISM.
However, we note that basic information such as age or gender is missing.

We can create some new columns with the available information so that they can be used for downstream analyses.
We will also make sure that they are encoded as factors and that the correct reference layer is used:

```{r}
# disease status
rse$disease_status <- sapply(rse$characteristics, "[", 1)
rse$disease_status <- sub("disease status: ", "", rse$disease_status)
rse$disease_status <- sub("systemic lupus erythematosus \\(SLE\\)", "SLE", rse$disease_status)
rse$disease_status <- factor(rse$disease_status, levels = c("healthy", "SLE"))
# tissue
rse$tissue <- sapply(rse$characteristics, "[", 2)
rse$tissue <- sub("tissue: ", "", rse$tissue)
rse$tissue <- factor(rse$tissue)
# anti-ro
rse$anti_ro <- sapply(rse$characteristics, "[", 3)
rse$anti_ro <- sub("anti-ro: ", "", rse$anti_ro)
rse$anti_ro <- factor(rse$anti_ro)
# ism
rse$ism <- sapply(rse$characteristics, "[", 4)
rse$ism <-sub("ism: ", "", rse$ism)
rse$ism <- factor(rse$ism)
```

We can check how many samples we have in each group (note that we ignore `tissue` as it's always `whole blood`):

```{r}
metadata <- data.frame(disease_status = rse$disease_status, anti_ro.ism = paste(rse$anti_ro, rse$ism, sep = "."))
table(metadata)
```

Now we are ready to perform a simple differential gene expression analysis with `DESeq2` [@Love2014].
Note that we remove genes with a low number of counts (less than 50 across all 117 samples) to speed up execution and reduce the memory footprint:

```{r}
library(DESeq2)
dds <- DESeqDataSet(rse, ~ disease_status)
dds <- DESeq(dds)
dds <- dds[rowSums(counts(dds)) >= 50, ]
```

We used an extremely simple model; in the real world we should be accounting for co-variables, potential confounders and interactions between them.
For example, age and gender are usually included in this type of analysis, but we don't have access to this information for this dataset.
Similarly, the value of the ISM and the presence of anti-Ro autoantibodies can't be included in the analysis due to the fact that these variables are collinear with the disease status variable (i.e.: the value of both `anti_ro` and `ism` is `control` for all samples with `disease_status` equal to `healthy`.)
Like `DESeq2`, `edgeR` [@Robinson2010] and `limma` [@Ritchie2015] can also deal with multiple cofactors and different experimental designs, and constitute good alternatives for performing differential expression analyses.

We can now look at the data in more detail to assess if we can observe a separation between the SLE and healthy samples and whether any batch effect is visible.  
We use the variance stabilising transformation (VST) [@Anders2010] for visualisation purposes:

```{r}
vsd <- vst(dds, blind = FALSE)
```

We will use the [`pheatmap`](http://cran.r-project.org/package=pheatmap) and [`RColorBrewer`](https://cran.r-project.org/package=RColorBrewer) packages to perform hierarchical clustering of the samples (Figure 2):

```{r figure2, fig.cap = 'Heatmap showing Euclidean distances between samples clustered using complete linkage. Disease status and other experimental factors are visualised as column annotations.'}
library(pheatmap)
library(RColorBrewer)
sampleDists <- dist(t(assay(vsd)))
sampleDistMatrix <- as.matrix(sampleDists)
annotation = data.frame(colData(vsd)[c("anti_ro", "ism", "disease_status")], row.names = rownames(sampleDistMatrix))
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pheatmap(sampleDistMatrix, clustering_distance_rows = sampleDists, clustering_distance_cols = sampleDists, clustering_method = "complete", annotation_col = annotation, col = colors, show_rownames = FALSE, show_colnames = FALSE, cellwidth = 2, cellheight = 2)
```

While there isn't an unambiguous split between healthy and disease samples, the most distinct clusters (bottom right and top left) are entirely composed of SLE samples, with the central cluster containing all healthy samples and a number of SLE ones.
The clusters don't appear to be due to the ISM or the presence of anti-Ro autoantibodies.

Similarly, we can perform a principal component analysis (PCA) on the most variable 500 genes (Figure 3).
Note that we load `ggplot2` [@Wickham2009] to modify the look of the plot:

```{r figure3, fig.cap = 'Scatter plot showing results of a PCA with samples coloured according to their disease status.'}
library(ggplot2)
plotPCA(vsd, intgroup = "disease_status") +
  coord_fixed()
```

We can see some separation of healthy and SLE samples along both PC1 and PC2, though some SLE samples appear very similar to the healthy ones.
No obvious batch effects are visible from this plot.

Next, we select genes that are differentially expressed below a 0.05 adjusted *p*-value threshold:

```{r}
res <- results(dds, alpha = 0.05)
summary(res)
```

We can visualise the shrunken log2 fold changes using an MA plot (Figure 4):

```{r figure4, fig.cap = 'MA plot showing genes differentially expressed (red dots) in SLE patients compared to healthy patients.'}
res_lfc <- lfcShrink(dds, coef = 2)
plotMA(res_lfc, ylim = c(-3, 3))
```

We observe large numbers of genes differentially expressed in both directions and across a range of fold changes, though the majority of significant genes appear to be upregulated in disease.

For convenience, we will save our differentially expressed genes (DEGs) in another object and map the GENCODE gene IDs to gene symbols using the annotation in the original `RangedSummarizedExperiment` object

```{r}
degs <- subset(res, padj < 0.05)
degs <- merge(rowData(rse), as.data.frame(degs), by.x = "gene_id", by.y = "row.names", all = FALSE)
head(degs, 3)
```

## Accessing GWAS data

The differential expression analysis resulted in several thousands of DEGs.
Since we know that genes with high levels of differential expression are more likely to harbour disease-associated variants [@Chen2008] and that therapeutic targets with genetic evidence are more likely to progress through the drug discovery pipeline [@Nelson2015], one way to prioritise them is to check which of these can be genetically linked to SLE.
To get hold of relevant GWAS data, we will be using the `gwascat` Bioconductor package [@Carey2017a], which provides an interface to the GWAS catalog [@MacArthur2017].
An alternative is to use the GRASP [@Eicher2015] database with the `grasp2db` [@Carey2017b] package.

```{r}
library(gwascat)
# uncomment the following line to download file and build the gwasloc object all in one step
#snps <- makeCurrentGwascat()
# uncomment the following line to download file
#download.file("http://www.ebi.ac.uk/gwas/api/search/downloads/alternative", destfile = "gwas_catalog_v1.0.1-associations_e90_r2017-12-04.tsv")
snps <- read.delim("gwas_catalog_v1.0.1-associations_e90_r2017-12-04.tsv", check.names = FALSE, stringsAsFactors = FALSE)
snps <- gwascat:::gwdf2GRanges(snps, extractDate = "2017-12-04")
genome(snps) <- "GRCh38"
head(snps, 3)
```

`snps` is a `gwasloc` object which is simply a wrapper around a `GRanges` object, the standard way to represent genomic ranges in Bioconductor.

We note here that the GWAS catalog uses GRCh38 coordinates, the same assembly used in the GENCODE v25 annotation. 
When integrating genomic datasets from different sources it is essential to ensure that the same genome assembly is used, especially because many datasets in the public domain are still using GRCh37 coordinates.
As we will see below, it is possible and relatively straightforward to convert genomic coordinates between genome assemblies.

We can select only SNPs that are associated with SLE:

```{r}
snps <- subsetByTraits(snps, tr = "Systemic lupus erythematosus")
```

We can visualise these as a Manhattan plot to look at the distribution of GWAS *p*-values over chromosomes on a negative log~10~ scale (Figure 5):
Note that *p*-values lower than 1e-25 are truncated in the figure:

```{r figure5, fig.cap = 'Manhattan plot showing GWAS variants significantly associated with SLE.'}
traitsManh(gwr = snps, sel = snps, traits = "Systemic lupus erythematosus") +
  xlab("SLE GWAS SNPs") +
  ylab("-log10(p-value)") +
  theme(legend.position = "none",
        strip.text.x = element_text(size = 6),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
```

We observe several hits across most chromosomes, with many of them below a genome-wide significant threshold (*p*-value < 1 x 10^-8^), suggesting that genetics plays an important role in the pathogenesis of SLE.

We note here that genotyping arrays typically include a very small fraction of all possible SNPs in the human genome, and there is no guarantee that the *tag* SNPs on the array are the true casual SNPs [@Bush2012].
The alleles of other SNPs can be imputed from tag SNPs thanks to the structure of linkage disequilibrium (LD) blocks present in chromosomes. 
Thus, when linking variants to target genes in a real-world setting, it is important to take into consideration neighbouring SNPs that are in high LD (e.g.: r^2^ > 0.8) and inherited with the tag SNPs.
Unfortunately, at the time of writing there is no straightforward way to perform this LD expansion step using R or Bioconductor packages, possibly because of the large amount of reference data required.
The `ldblock` package [@Carey2017c] used to provide this functionality by downloading the HapMap data from the NCBI website, but the dataset was retired in 2016.
At present, the best option to do this programmatically is probably to query the Ensembl REST API [@Yates2015].

## Annotation of coding and proximal SNPs to target genes

In order to annotate these variants, we need a a `TxDb` object, a reference of where transcripts are located on the genome.
We can build this using the `GenomicFeatures` [@Lawrence2013] package and the GENCODE v25 gene annotation:

```{r}
library(GenomicFeatures)
# uncomment the following line to download file
#download.file("ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_25/gencode.v25.annotation.gff3.gz", destfile = "gencode.v25.annotation.gff3.gz")
txdb <- makeTxDbFromGFF("gencode.v25.annotation.gff3.gz")
txdb <- keepStandardChromosomes(txdb)
```

We also have to convert the `gwasloc` object into a standard `GRanges` object:

```{r}
snps <- GRanges(snps)
```

Let's check if the `gwasloc` and `TxDb` object use the same notation for chromosomes:

```{r}
seqlevelsStyle(snps)
seqlevelsStyle(txdb)
```

OK, they do.
Now we can annotate our SNPs to genes using the `VariantAnnotation` [@Obenchain2014] package:

```{r}
library(VariantAnnotation)
snps_anno <- locateVariants(snps, txdb, AllVariants())
snps_anno <- unique(snps_anno)
```

We use the `QUERYID` column in `snps_anno` to recover metadata such as SNP IDs and GWAS *p*-values from the original `snps` object:

```{r}
snps_metadata <- snps[snps_anno$QUERYID]
mcols(snps_anno) <- cbind(mcols(snps_metadata)[c("SNPS", "P-VALUE")], mcols(snps_anno))
```

We can visualise where these SNPs are located (Figure 6):

```{r figure6, fig.cap = 'Bar plot showing genomic locations associated with SLE variants.'}
loc <- data.frame(table(snps_anno$LOCATION))
ggplot(data = loc, aes(x = reorder(Var1, -Freq), y = Freq)) +
  geom_bar(stat = "identity") +
  xlab("Genomic location of SNPs") +
  ylab("Number of SNPs")
```

As expected [@Maurano2012], the great majority of SNPs are located within introns and in intergenic regions.
For the moment, we will focus on SNPs that are either coding or in promoter and UTR regions, as these can be assigned to target genes rather unambiguously:

```{r}
snps_easy <- subset(snps_anno, LOCATION == "coding" | LOCATION == "promoter" | LOCATION == "threeUTR" | LOCATION == "fiveUTR")
snps_easy <- as.data.frame(snps_easy)
```

Now we can check if any of the genes we found to be differentially expressed in SLE is also genetically associated with the disease:

```{r}
snps_easy_in_degs <- merge(degs, snps_easy, by.x = "gene_id", by.y = "GENEID", all = FALSE)
```

We have `r length(unique(snps_easy_in_degs$gene_id))` genes showing differential expression in SLE that are also genetically associated with the disease.
While this is an interesting result, these hits are likely to be already well-known as potential SLE targets given their clear genetic association.

We will store essential information about these hits in a results `data.frame`:

```{r}
prioritised_hits <- unique(data.frame(
  snp_id = snps_easy_in_degs$SNPS,
  snp_pvalue = snps_easy_in_degs$P.VALUE,
  snp_location = snps_easy_in_degs$LOCATION,
  gene_id = snps_easy_in_degs$gene_id,
  gene_symbol = snps_easy_in_degs$symbol,
  gene_pvalue = snps_easy_in_degs$padj,
  gene_log2foldchange = snps_easy_in_degs$log2FoldChange,
  method = "Direct overlap",
  row.names = NULL))
head(prioritised_hits, 3)
```

## Use of regulatory genomic data to map intronic and intergenic SNPs to target genes

But what about all the SNPs in introns and intergenic regions?
Some of those might be regulatory variants affecting the expression level of their target gene(s) through a distal enhancer.
Let's create a dataset of candidate regulatory SNPs that are either intronic or intergenic and remove the annotation obtained with `VariantAnnotation`:

```{r}
snps_hard <- subset(snps_anno, LOCATION == "intron" | LOCATION == "intergenic", select = c("SNPS", "P.VALUE", "LOCATION"))
```

### eQTL data

A well-established way to gain insights into target genes of regulatory SNPs is to use eQTL data, where correlations between genetic variants and expression of genes are computed across different tissues or cell types [@Albert2015].
Here, we will simply match GWAS SNPs and eQTLs according to their genomic locations, which is a rather crude way to integrate these two types of data.
More robust alternatives such as PrediXcan [@Gamazon2015], TWAS [@Gusev2016] and SMR [@Zhu2016] exist and should be adopted if possible.
One downside of these methods is that they require subject-level or complete summary data, making them less practical in some circumstances.

We will use blood eQTL data from the GTEx consortium [@GTEx2017].
To get the data, you will have to register and download the file `GTEx_Analysis_v7_eQTL.tar.gz` from the [GTEx portal](https://www.gtexportal.org) to the current working directory:

```{r}
# uncomment the following line to extract the gzipped archive file
#untar("GTEx_Analysis_v7_eQTL.tar.gz")
gtex_blood <- read.delim(gzfile("GTEx_Analysis_v7_eQTL/Whole_Blood.v7.signif_variant_gene_pairs.txt.gz"), stringsAsFactors = FALSE)
head(gtex_blood, 3)
```

We have to extract the genomic locations of the SNPs from the IDs used by GTEx:

```{r}
locs <- strsplit(gtex_blood$variant_id, "_")
gtex_blood$chr <- sapply(locs, "[", 1)
gtex_blood$start <- sapply(locs, "[", 2)
gtex_blood$end <- sapply(locs, "[", 2)
```

We can then convert the `data.frame` into a `GRanges` object:

```{r}
gtex_blood <- makeGRangesFromDataFrame(gtex_blood, keep.extra.columns = TRUE)
```

We also need to ensure that the chromosome notation is consistent with the previous objects:

```{r}
seqlevelsStyle(gtex_blood)
seqlevelsStyle(gtex_blood) <- "UCSC"
```

From the publication [@GTEx2017], we know the genomic coordinates are mapped to genome reference GRCh37, so we will have to uplift them to GRCh38 using `rtracklayer` [@Lawrence2009] and a mapping ("chain") file.
The [`R.utils`](https://cran.r-project.org/package=R.utils) package is only required to extract the gzipped file: 

```{r}
library(rtracklayer)
library(R.utils)
# uncomment the following line to download file
#download.file("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg38.over.chain.gz", destfile = "hg19ToHg38.over.chain.gz")
# uncomment the following line to extract gzipped file
#gunzip("hg19ToHg38.over.chain.gz")
ch <- import.chain("hg19ToHg38.over.chain")
gtex_blood <- unlist(liftOver(gtex_blood, ch))
```

We will use the `GenomicRanges` package [@Lawrence2013] to compute the overlap between GWAS SNPs and blood eQTLs:

```{r}
library(GenomicRanges)
hits <- findOverlaps(snps_hard, gtex_blood)
snps_hard_in_gtex_blood = snps_hard[queryHits(hits)]
gtex_blood_with_snps_hard = gtex_blood[subjectHits(hits)]
mcols(snps_hard_in_gtex_blood) <- cbind(mcols(snps_hard_in_gtex_blood), mcols(gtex_blood_with_snps_hard))
snps_hard_in_gtex_blood <- as.data.frame(snps_hard_in_gtex_blood)
```

We have `r length(unique(snps_hard_in_gtex_blood$SNPS))` blood eQTL variants that are associated with SLE.
We can now check whether any of the genes differentially expressed in SLE is an *eGene*, a gene whose expression is influenced by an eQTL.
Note that gene IDs in GTEx are mapped to GENCODE v19 [@GTEx2017], while we are using the newer v25 for the DEGs.
To match the gene IDs in the two objects, we will simply strip the last bit containing the GENCODE gene version, which effectively gives us Ensembl gene IDs:

```{r}
snps_hard_in_gtex_blood$ensembl_id <- sub("(ENSG[0-9]+)\\.[0-9]+", "\\1", snps_hard_in_gtex_blood$gene_id)
degs$ensembl_id <- sub("(ENSG[0-9]+)\\.[0-9]+", "\\1", degs$gene_id)
snps_hard_in_gtex_blood_in_degs <- merge(snps_hard_in_gtex_blood, degs, by = "ensembl_id", all = FALSE)
```

We can add these `r length(unique(snps_hard_in_gtex_blood_in_degs$gene_id.y))` genes to our list:

```{r}
prioritised_hits <- unique(rbind(prioritised_hits, data.frame(
  snp_id = snps_hard_in_gtex_blood_in_degs$SNPS,
  snp_pvalue = snps_hard_in_gtex_blood_in_degs$P.VALUE,
  snp_location = snps_hard_in_gtex_blood_in_degs$LOCATION,
  gene_id = snps_hard_in_gtex_blood_in_degs$gene_id.y,
  gene_symbol = snps_hard_in_gtex_blood_in_degs$symbol,
  gene_pvalue = snps_hard_in_gtex_blood_in_degs$padj,
  gene_log2foldchange = snps_hard_in_gtex_blood_in_degs$log2FoldChange,
  method = "GTEx eQTLs",
  row.names = NULL)))
```

### FANTOM5 data

The FANTOM consortium profiled gene expression across a large panel of tissues and cell types using CAGE [@Fantom2014; @Andersson2014].
This technology allows mapping of transcription start sites and enhancer RNAs genome-wide.
Correlations between these promoter and enhancer elements across a large panel of tissues and cell types can then be calculated to identify significant promoter - enhancer pairs.
In turn, we will use these correlations to map distal regulatory SNPs to target genes.

Let's read in the enhancer - promoter correlation data:

```{r}
# uncomment the following line to download the file
#download.file("http://enhancer.binf.ku.dk/presets/enhancer_tss_associations.bed", destfile = "enhancer_tss_associations.bed")
fantom <- read.delim("enhancer_tss_associations.bed", skip = 1, stringsAsFactors = FALSE)
head(fantom, 3)
```

Everything we need is in the fourth column, `name`: genomic location of the enhancer, gene identifiers, Pearson correlation coefficient and significance.
We will use the [`splitstackshape`](https://cran.r-project.org/package=splitstackshape) package to parse it:

```{r}
library(splitstackshape)
fantom <- as.data.frame(cSplit(fantom, splitCols = "name", sep = ";", direction = "wide"))
```

Now we can extract the genomic locations of the enhancers and the correlation values:

```{r}
locs <- strsplit(as.character(fantom$name_1), "[:-]")
fantom$chr <- sapply(locs, "[", 1)
fantom$start <- as.numeric(sapply(locs, "[", 2))
fantom$end <- as.numeric(sapply(locs, "[", 3))
fantom$symbol <- fantom$name_3
fantom$corr <- sub("R:", "", fantom$name_4)
fantom$fdr <- sub("FDR:", "", fantom$name_5)
```

We can select only the enhancer - promoter pairs with a decent level of correlation and significance and tidy the data at the same time:

```{r}
fantom <- unique(subset(fantom, corr >= 0.25 & fdr < 1e-5, select = c("chr", "start", "end", "symbol")))
```

Now we would like to check whether any of our candidate regulatory SNPs are falling in any of these enhancers.
To do this, we have to convert the `data.frame` into a `GRanges` object and uplift the GRCh37 coordinates [@Fantom2014] to GRCh38:

```{r}
fantom <- makeGRangesFromDataFrame(fantom, keep.extra.columns = TRUE)
fantom <- unlist(liftOver(fantom, ch))
```

We can now compute the overlap between SNPs and enhancers:

```{r}
hits <- findOverlaps(snps_hard, fantom)
snps_hard_in_fantom = snps_hard[queryHits(hits)]
fantom_with_snps_hard = fantom[subjectHits(hits)]
mcols(snps_hard_in_fantom) <- cbind(mcols(snps_hard_in_fantom), mcols(fantom_with_snps_hard))
snps_hard_in_fantom <- as.data.frame(snps_hard_in_fantom)
```

We can now check if any of these genes is differentially expressed in our RNA-seq data:

```{r}
snps_hard_in_fantom_in_degs <- merge(snps_hard_in_fantom, degs, by = "symbol", all = FALSE)
```

We have identified `r length(unique(snps_hard_in_fantom_in_degs$gene_id))` genes whose putative enhancers contain SLE GWAS SNPs.
Let's add these to our list:

```{r}
prioritised_hits <- unique(rbind(prioritised_hits, data.frame(
  snp_id = snps_hard_in_fantom_in_degs$SNPS,
  snp_pvalue = snps_hard_in_fantom_in_degs$P.VALUE,
  snp_location = snps_hard_in_fantom_in_degs$LOCATION,
  gene_id = snps_hard_in_fantom_in_degs$gene_id,
  gene_symbol = snps_hard_in_fantom_in_degs$symbol,
  gene_pvalue = snps_hard_in_fantom_in_degs$padj,
  gene_log2foldchange = snps_hard_in_fantom_in_degs$log2FoldChange,
  method = "FANTOM5 correlations",
  row.names = NULL)))
```

### Promoter Capture Hi-C data

More recently, chromatin interaction data was generated across 17 human primary blood cell types using promoter capture Hi-C [@Javierre2016].
More than 30,000 promoter baits were used to capture promoter-interacting regions genome-wide, which were then mapped to enhancers based on annotation present in the Ensembl Regulatory Build [@Zerbino2015].
This dataset provides a valuable resource for interpreting complex genomic data, especially in the context of autoimmune diseases (and other conditions where immune cells play a role). 
Significant interactions between enhancers and promoters can be accessed in the supplementary data of the paper:

```{r}
# uncomment the following line to download file
#download.file("http://www.cell.com/cms/attachment/2086554122/2074217047/mmc4.zip", destfile = "mmc4.zip")
# uncomment the following lines to extract zipped files
#unzip("mmc4.zip")
#unzip("DATA_S1.zip")
pchic <- read.delim("ActivePromoterEnhancerLinks.tsv", stringsAsFactors = FALSE)
head(pchic, 3)
```

We will use the `InteractionSet` package [@Lun2016], which is specifically designed for the representation of chromatin interaction data.
We start by creating a `GInteractions` object:

```{r}
library(InteractionSet)
promoters <- GRanges(seqnames = pchic$baitChr, ranges = IRanges(start = pchic$baitSt, end = pchic$baitEnd))
enhancers <- GRanges(seqnames = pchic$oeChr, ranges = IRanges(start = pchic$oeSt, end = pchic$oeEnd))
pchic <- GInteractions(promoters, enhancers)
```

As gene identifiers are not provided, we also have to map promoters to the respective genes so that we know which genes are regulated by which enhancers.
We can do this by using the `TxDb` object we previously built to extract positions of transcription start sites (TSSs) and then add the GENCODE gene IDs as metadata to the `pchic` object:

```{r}
tsss <- promoters(txdb, upstream = 0, downstream = 1, columns = "gene_id")
hits <- nearest(promoters, tsss)
pchic$gene_id <- unlist(tsss[hits]$gene_id)
```

Next, we calculate the overlaps between SLE GWAS SNPs and enhancers (the *second* region of the `GInteractions` object) :

```{r}
hits <- findOverlaps(snps_hard, pchic, use.region = "second")
snps_hard_in_pchic = snps_hard[queryHits(hits)]
pchic_with_snps_hard = pchic[subjectHits(hits)]
mcols(snps_hard_in_pchic) <- cbind(mcols(snps_hard_in_pchic), mcols(pchic_with_snps_hard))
snps_hard_in_pchic <- as.data.frame(snps_hard_in_pchic)
```

We check if any of these enhancers containing SLE variants are known to putatively regulate genes differentially expressed in SLE:

```{r}
snps_hard_in_pchic_in_degs <- merge(snps_hard_in_pchic, degs, by = "gene_id", all = FALSE)
```

And finally we add these `r length(unique(snps_hard_in_pchic_in_degs$gene_id))` genes to our list:

```{r}
prioritised_hits <- unique(rbind(prioritised_hits, data.frame(
  snp_id = snps_hard_in_pchic_in_degs$SNPS,
  snp_pvalue = snps_hard_in_pchic_in_degs$P.VALUE,
  snp_location = snps_hard_in_pchic_in_degs$LOCATION,
  gene_id = snps_hard_in_pchic_in_degs$gene_id,
  gene_symbol = snps_hard_in_pchic_in_degs$symbol,
  gene_pvalue = snps_hard_in_pchic_in_degs$padj,
  gene_log2foldchange = snps_hard_in_pchic_in_degs$log2FoldChange,
  method = "Promoter capture Hi-C",
  row.names = NULL)))
```

These are the final results of our target identification exercise.
We can have a look at the most significant SNPs mapped with each of the methods:

```{r}
top_prioritised_hits <- prioritised_hits[order(prioritised_hits$snp_pvalue),]
top_prioritised_hits <- split(top_prioritised_hits, top_prioritised_hits$method)
do.call(rbind, lapply(top_prioritised_hits, head, 1))
```

We can also visualise the relative contributions from the different approaches we used (Figure 7):

```{r figure7, fig.cap = 'Bar plot showing number of genes identified by each variant mapping method.'}
prioritised_genes <- unique(data.frame(gene_id = prioritised_hits$gene_id, method = prioritised_hits$method))
ggplot(data = prioritised_genes, aes(x = method)) +
  geom_bar(aes(fill = method), stat = "count") +
  ylab("Number of genes") +
  theme(axis.title.x = element_blank(),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank())
```

We observe that all methods significantly contributed to the identification of genes associated with GWAS SNPs.
The majority of genes were identified through the integration of the GTEx blood eQTL data, followed by the methods based on direct overlap, promoter capture Hi-C data and FANTOM5 correlations.

## Functional analysis of priortised hits

We will use biological processes from the Gene Ontology [@Ashburner2000] and the `clusterProfiler` package [@Yu2012] to functionally characterise our list of genes:

```{r}
library(clusterProfiler)
prioritised_hits_ensembl_ids <- unique(sub("(ENSG[0-9]+)\\.[0-9]+", "\\1", prioritised_hits$gene_id))
all_genes_ensembl_ids <- unique(sub("(ENSG[0-9]+)\\.[0-9]+", "\\1", rownames(rse)))
gobp_enrichment <- enrichGO(prioritised_hits_ensembl_ids,
                            universe = all_genes_ensembl_ids,
                            OrgDb = org.Hs.eg.db,
                            keyType = "ENSEMBL",
                            ont = "BP",
                            pAdjustMethod = "BH",
                            pvalueCutoff = 0.05,
                            qvalueCutoff = 0.05,
                            readable = TRUE)
```

We can visualise the most enriched terms (Figure 8):

```{r figure8, fig.cap = 'Dot plot showing enrichment of Gene Ontology biological processes for the list of prioritised genes.', fig.width = 10}
dotplot(gobp_enrichment, showCategory = 20)
```

We observe a significant enrichment for interferon responses, antigen processing and presentation, and T cell stimulation, all processes which are well-known to play key roles in the pathogenesis of SLE [@Oon2016;@Morris2014;@Suarez-Fueyo2016].

From a drug discovery perspective, JAK2 is probably the most attractive target: rs1887428 (*p*-value = 1 x 10^-6^) is located in its 5' UTR and the genes is significantly upregulated in disease.
Tofacitinib, a pan-JAK inhibitor, showed promising results in mouse [@Furumoto2017] and is currently being tested or safety in a [phase I clinical trial](https://clinicaltrials.gov/ct2/show/NCT02535689).
We find 7 GWAS SNPs that are blood eQTLs linked to the expression of C2, a protease active in the complement signalling cascade.
The most significant variant is rs1270942 (*p*-value = 2 x 10^-165^) and is found in an intron of CFB, another component of the complement system.
As with other autoimmune diseases, the complement plays a key role SLE in and has been investigated as a therapeutic approach [@Leffler2014].
Another potentially interesting hit is TAX1BP1: rs849142 (*p*-value = 1 x 9^-11^) is found within an intron of JAZF1, but can be linked to TAX1BP1 via a chromatin interaction with its promoter.
TAX1BP1 inhibits TNF-induced apoptosis [@DeValck1999] and is involved in the IL1 signalling cascade [@Ling2000], another relevant pathway in SLE that could be therapeutically targeted [@Ronnblom2010].

# Conclusions

In this Bioconductor workflow we have used several packages and datasets to demonstrate how regulatory genomic data can be used to annotate significant hits from GWASs and prioritise gene lists from expression studies, providing an intermediate layer connecting genetics and transcriptomics.
Overall, we identified `r length(unique(prioritised_hits$snp_id))` SLE-associated SNPs that we mapped to `r length(unique(prioritised_hits$gene_id))` genes differentially expressed in SLE, using eQTL data [@GTEx2017] and enhancer - promoter relationships from CAGE [@Fantom2014] and promoter capture Hi-C experiments [@Javierre2016].
These genes are involved in key inflammatory signalling pathways and some of them could develop into therapeutic targets for SLE.
While options for the visualisations of genomic data and interactions are outside the scope of this workflow, at least three good alternatives exist in Bioconductor: ggbio [@Yin2012], Sushi [@Phanstiel2014] and Gviz [@Hahne2016] coupled with the GenomicInteractions package [@Harmston2015].
We refer the reader to these publications and package vignettes for examples.

While simplified, the workflow also demonstrates some real-world challenges encountered when working with genomic data from different sources, such as the use of different genome assemblies and gene annotation systems, the parsing of files with custom formats into Bioconductor objects and the mapping of genomic locations to genes.

As the sample size and power of GWASs and gene expression studies continue to increase, it will become more and more challenging to identify truly significant hits and interpret them.
The use of regulatory genomics data as presented here can be an important tool to gain insights into large biomedical datasets and help in the identification of biomarkers and therapeutic targets. 

# Abbreviations

CAGE: cap analysis of gene expression  
DHS: DNase I hypersensitive site  
eQTL: expression quantitative trait locus  
GWAS: genome-wide association study  
ISM: interferon signature metric  
SLE: systemic lupus erythematosus  
SNP: single nucleotide polymorphism  

# Data and software availability

Download links for all datasets are part of the workflow.  
Software packages required to reproduce the analysis can be installed as part of the workflow.  
Source code is available at: https://github.com/enricoferrero/bioconductor-regulatory-genomics-workflow.  
Archived source code as at the time of publication is available at: https://doi.org/10.5281/zenodo.1154124.  

# Competing interests

EF is a full time employee of GSK.

# Grant information

The author declared that no grants were involved in supporting this work.

# References
