---
title: "Bioconductor Regulatory Genomics Workflow"
author:
  name: Enrico Ferrero
  email: enrico.x.ferrero@gsk.com
  affiliation: Computational Biology, GSK, Medicines Research Centre, Gunnels Wood Road, Stevenage, SG1 2NY, UK
abstract: The identification of therapeutic targets is a critical issue in drug discovery, with several programmes failing because of a weak linkage between target and disease. Genome-wide association studies and large gene expression experiments are providing insights into the biology of several common and complex diseases, but the complexity of transcriptional regulation mechanisms often limit our understanding of how genetic variation can influence changes in gene expression. Several initiatives in the field of regulatory genomics are aiming to close this gap by systematically identifying and cataloguing regulatory elements such as promoters and enhacersacross different tissues and cell types. In this Bioconductor workflow, we will explore how different types of regulatory genomic data can be used for the functional interpretation of disease-associated variants and for the prioritisation of gene lists from gene expression experiments.
keywords: bioconductor; r; rstats; regulatory genomics; functional genomics; genetics; gwas; transcriptomics; integration; multiomics
output:
  html_document:
    theme: flatly
    highlight: tango
  github_document:
    html_preview: false
  BiocWorkflowTools::f1000_article: default
bibliography: references.bib
biblio-style: unsrtnat
natbiboptions: super
csl: biomed-central.csl
editor_options: 
  chunk_output_type: console
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(message = FALSE, warning = FALSE, fig.width = 6, fig.height = 6)
```

# Introduction

Discovering and bringing new drugs to the market is a long, expensive and inefficient process [@Waring2015; @DiMasi2016].
Increasing the success rates of drug discovery programmes would be transformative to the pharmaceutical industry and significantly improve patientsâ€™ access to medicines.
Of note, the majority of drug discovery programmes fail for efficacy reasons [@Harrison2016], with up to 40% of these failures due to lack of a clear link between the target and the disease under investigation [@Cook2014].

Target selection, the first step in drug discovery programmes, is a critical decision point.
It has previously been shown that therapeutic targets with a genetic link to the disease under investigation are more likely to progress through the drug discovery pipeline, suggesting that genetics can be used as a tool to prioritise and validate drug targets in early discovery [@Plenge2013; @Nelson2015].

Over the last decade, genome-wide association studies (GWASs) have revolutionised the field of human genetics allowing to survey DNA mutations associated with disease and other complex traits on an unprecedented scale [@Visscher2017].
Similarly, phenome-wide association studies (PheWAS) are emerging as a complementary methodology to decipher the genetic bases of the human phenome [@Bush2016].
While many of these associations might not actually be relevant for the disease aetiology [@Boyle2017], these methods hold much promise to guide towards the next generation of drug targets [@Finan2017]. 

Arguably, the biggest challenge in translating findings from GWASs to therapies is that the great majority of single nucleotide polymorphisms (SNPs) associated with disease are found in non-coding regions of the genome and therefore cannot be easily linked to a target gene [@Maurano2012].
Many of these SNPs could be regulatory variants, affecting the expression of nearby or distal genes by interfering with the process of transcription (e.g.: binding of transcription factors at promoters or enhancers) [@Ward2012].

The most established way to map disease-associated regulatory variants to target genes is probably to use expression quantitative trait loci (eQTLs) [@Albert2015], variants that affect the expression of specific genes.
Over the last few years, the GTEx consortium assembled a valuable resource by performing large-scale mapping of genome-wide correlations between genetic variants and gene expression across 44 human tissues [@GTEx2017a].

However, depending on the power of the study, it might not be possible to detect all existing regulatory variants as eQTLs.
An alternative is to use information on the location of promoters and distal enhancers across the genome and link these regulatory elements to their target genes.
Large, multi-centre Initiatives such as ENCODE [@ENCODE2012], Roadmap Epigenomics [@Roadmap2015] and BLUEPRINT [@Adams2012; @Stunnenberg2016] mapped regulatory elements in the genome by profiling a number of chromatin features including DNase hypersensitive sites (DHSs), several types of histone marks and binding of chromatin-associated proteins in a large number of cell lines, primary cell types and tissues.
Similarly, the FANTOM consortium used cap analysis of gene expression (CAGE) to identify promoters and enhancers across hundreds of cells and tissues [@Fantom2014].

Knowing that a certain stretch of DNA is an enhancer is however not informative of the target gene(s).
One way to infer links between enhancers and promoters *in silico* is to identify significant correlations across a large panel of cell types, an approach that was used for distal and promoter DHSs [@Thurman2012] as well as for CAGE-defined promoters and enhancers [@Andersson2014].
Experimental methods to assay interactions between regulatory elements also exist.
Chromatin interaction analysis by paired-end tag sequencing (ChIA-PET) [@Fullwood2009; @Zhang2013] couples chromatin immunoprecipitation with DNA ligation and sequencing to identify regions of DNA that are interacting thanks to the binding of a specific protein.
Promoter capture Hi-C [@Mifsud2015; @Javierre2016] extends chromatin conformation capture by using "baits" to enrich for promoter interactions and increase resolution.

Overall, linking genetic variants to their candidate target genes is not straightforward, not only because of the complexity of the human genome and transcriptional regulation, but also because of the variety of data types and approaches that can be used.
To address this, we developed STOPGAP (systematic target opportunity assessment by genetic association predictions), a database of disease variants mapped to their most likely target gene(s) using different types of regulatory genomic data [@Shen2017].
The database is currently undergoing a major overhaul and will eventually be superseded by POSTGAP [@Ensembl2017].
A similar resource and valid alternative is INFERNO (inferring the molecular mechanisms of noncoding variants) [@Amlie-Wolf2017].

# Workflow

## Overview

In this workflow we will explore how regulatory genomic data can be used to connect the genetic and transcriptional layers by providing a framework for the functional annotation of SNPs from GWASs.
We will use eQTL data from GTEx [@GTEx2017a], FANTOM5 correlations between promoters and enhancers [@Andersson2014] and promoter capture Hi-C data [Javierre2016].

We start with a common scenario: we run a RNA-seq experiment comparing patients with a disease and healthy individuals, and would like to discover key disease genes and potential therapeutic targets by integrating genetic information in our analysis.

## Install required packages

R version 3.4.2 and Bioconductor version 3.6 were used for the analysis.
The code below will install all required packages and dependencies from Bioconductor and CRAN.

```{r}
source("https://bioconductor.org/biocLite.R")
# uncomment following line to install packages
#biocLite(c("DESeq2", "GenomicFeatures", "GenomicRanges", "ggplot2", "gwascat", "recount", "pheatmap", "RColorBrewer", "rtracklayer", "R.utils", "splitstackshape", "VariantAnnotation"))
```

## Gene expression data and differential gene expression analysis

The RNA-seq data we will be using comes from blood of patients with systemic lupus erythematosus (SLE) and healthy controls [@Hung2015].

We are going to use `recount` [@Collado-Torres2017] to obtain gene-level counts:

```{r}
library(recount)
# uncomment following line to download dataset
#download_study("SRP062966")
load(file.path("SRP062966", "rse_gene.RData"))
rse <- scale_counts(rse_gene)
rse
```

Other Bioconductor packages that can be used to access data from gene expression experiments directly in R are `GEOquery` [@Davis2007] and `ArrayExpress` [@Kauffmann2009]. 

So, we have `r ncol(rse)` samples.
This is what the data looks like:

```{r}
assay(rse)[1:10, 1:10]
```

We note this is a GENCODE v25 annotation, which will be useful later on.
Let's look at the metadata to check how we can split them between cases and controls:

```{r}
colData(rse)
```

The most interesting part of the metadata is contained in the `characteristics` column, which is a `CharacterList` object:

```{r}
colData(rse)$characteristics
```

Let's create some new columns with this information that can be used for the differential expression analysis.
We will also make sure that they are encoded as factors and that the correct reference layer is used:

```{r}
# disease status
colData(rse)$disease_status <- sapply(colData(rse)$characteristics, "[", 1)
colData(rse)$disease_status <- sub("disease status: ", "", colData(rse)$disease_status)
colData(rse)$disease_status <- sub("systemic lupus erythematosus \\(SLE\\)", "SLE", colData(rse)$disease_status)
colData(rse)$disease_status <- factor(colData(rse)$disease_status, levels = c("healthy", "SLE"))
# tissue
colData(rse)$tissue <- sapply(colData(rse)$characteristics, "[", 2)
colData(rse)$tissue <- sub("tissue: ", "", colData(rse)$tissue)
colData(rse)$tissue <- factor(colData(rse)$tissue)
# anti-ro
colData(rse)$anti_ro <- sapply(colData(rse)$characteristics, "[", 3)
colData(rse)$anti_ro <- sub("anti-ro: ", "", colData(rse)$anti_ro)
colData(rse)$anti_ro <- factor(colData(rse)$anti_ro)
# ism
colData(rse)$ism <- sapply(colData(rse)$characteristics, "[", 4)
colData(rse)$ism <-sub("ism: ", "", colData(rse)$ism)
colData(rse)$ism <- factor(colData(rse)$ism)
```

Let's check it's what we expect:

```{r}
colData(rse)[c("disease_status", "tissue", "anti_ro", "ism")]
```

OK, this looks more readable.
Let's check how many samples we have in each group:

```{r}
table(colData(rse)$disease_status)
```

To speed up code execution we will limit the number of SLE samples.
For simplicity, we select the first 18 (healthy) and the last 18 (SLE) samples from the original `RangedSummarizedExperiment` object:

```{r}
rse <- rse[, c(1:18, 82:99)]
```

Now we are ready to perform a simple differential gene expression analysis with `DESeq2` [@Love2014]:

```{r}
library(DESeq2)
dds <- DESeqDataSet(rse, ~ disease_status)
dds <- DESeq(dds)
dds
```

Note that we used an extremely simple model; in the real world you will probably need to account for co-variables, potential confounders and interactions between them.
`edgeR` [@Robinson2010] and `limma` [@Ritchie2015] are good alternatives to `DESEq2` for performing differential expression analyses.

We can now look at the data in more detail.
We use the variance stabilising transformation (VST) [@Anders2010] for visualisation purposes:

```{r}
vsd <- vst(dds, blind = FALSE)
```

First, let's look at distances between samples to see if we can recover a separation between SLE and healthy samples:

```{r}
sampleDists <- as.matrix(dist(t(assay(vsd))))
rownames(sampleDists) <- vsd$disease_status
sampleDists[c(1, 18, 19, 36), c(1, 18, 19, 36)]
```

We will use the `pheatmap` [@Kolde2015] and `RColorBrewer` [Neuwirth2014] packages for drawing the heatmap (Figure \@ref(fig:heatmap)):

```{r heatmap, fig.cap = 'Clustered heatmap showing distances between samples.'}
library(pheatmap)
library(RColorBrewer)
colors <- colorRampPalette(rev(brewer.pal(9, "Blues")))(255)
pheatmap(sampleDists, col = colors)
```

Similarly, we can perform a principal component analysis (PCA) on the most variable 500 genes (Figure \@ref(fig:pca)):

```{r pca, fig.cap = 'Principal component analysis with samples coloured according to their disease status.'}
plotPCA(vsd, intgroup = "disease_status")
```

This looks better, we can see some separation of healthy and SLE samples along both PC1 and PC2, though some SLE samples appear very similar to the healthy ones.
Next, we select genes that are differentially expressed below a 0.05 adjusted p-value threshold:

```{r}
res <- results(dds, alpha = 0.05)
res
```

We can look at a summary of the results:

```{r}
summary(res)
```

We can also visualise the log fold changes using an MA plot (Figure \@ref(fig:maplot)):

```{r maplot, fig.cap = 'MA plot showing genes differentially expressed in SLE patients compared to healthy patients.'}
plotMA(res, ylim = c(-5,5))
```

For convenience, we will save our differentially expressed genes (DEGs) in another object:

```{r}
degs <- subset(res, padj < 0.05)
degs <- as.data.frame(degs)
head(degs)
```

We also map the GENCODE gene IDs to gene symbols using the annotation in the original `RangedSummarizedExperiment` object, which is going to be convenient later on:

```{r}
rowData(rse)
degs <- merge(rowData(rse), degs, by.x = "gene_id", by.y = "row.names", all = FALSE)
tail(degs)
```

## Accessing GWAS data

We have more than 3500 genes of interest at this stage.
Since we know that therapeutic targets with genetic evidence are more likely to progress through the drug discovery pipeline [@Nelson2015], one way to prioritise them could be to check which of these can be genetically linked to SLE.
To get hold of relevant GWAS data, we will be using the `gwascat` Bioconductor package [@Carey2017a], which provides an interface to the GWAS catalog [@MacArthur2017].
An alternative is to use the GRASP [@Eicher2015] database with the `grasp2db` [@Carey2017b] package.

```{r}
library(gwascat)
# uncomment following line to download file and build the gwasloc object all in one step
#snps <- makeCurrentGwascat()
# uncomment following line to download file
#download.file("http://www.ebi.ac.uk/gwas/api/search/downloads/alternative", destfile = "gwas_catalog_v1.0.1-associations_e90_r2017-12-04.tsv")
snps <- read.delim("gwas_catalog_v1.0.1-associations_e90_r2017-12-04.tsv", check.names = FALSE, stringsAsFactors = FALSE)
snps <- gwascat:::gwdf2GRanges(snps, extractDate = "2017-12-04")
genome(snps) <- "GRCh38"
snps
```

SNPs is a `gwasloc` object which is simply a wrapper around a `GRanges` object, the standard way to express genomic ranges in Bioconductor.
We are interested in SNPs associated with SLE:

```{r}
snps <- subsetByTraits(snps, tr = "Systemic lupus erythematosus")
snps
```

We can visualise these as a Manhattan plot to look at the distribution of GWAS p-values over chromosomes on a negative log scale (Figure \@ref(fig:manhattan), note that p-values lower than 1e-25 are truncated):

```{r manhattan, fig.cap = 'Manhattan plot showing variants significantly associated with SLE.', fig.height = 8, fig.width = 12}
traitsManh(gwr = snps, sel = snps, traits = "Systemic lupus erythematosus")
```

We note here that genotyping arrays typically include a very small fraction of all possible SNPs in the human genome, and there is no guarantee that the *tag* SNPs on the array are the true casual SNPs [@Bush2012].
The alleles of other SNPs can be imputed from tag SNPs thanks to the structure of linkage disequilibrium (LD) blocks present in chromosomes. 
Thus, when linking variants to target genes in a real-world setting, it is important to take into consideration neighbouring SNPs that are in high LD and inherited with the tag SNPs.
For simplicity, We will skip this LD expansion step and refer the reader to the Ensembl REST API [@Yates2015; @Ensembl2017a], the Ensembl Linkage Disequilibrium Calculator [@Ensembl2017b] and the Bioconductor packages `trio` [@Schwender2015] and `ldblock` [@Carey2017c] to perform this task.

## Annotation of coding and proximal SNPs to target genes 

In order to annotate these variants, We need a a `TxDb` object, a reference of where transcripts are located on the genome.
We can build this using the `GenomicFeatutres` [@Lawrence2013] package and annotation from GENCODE (which is the gene annotation used in our RNA-seq experiment):

```{r}
library(GenomicFeatures)
#download.file("ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_25/gencode.v25.annotation.gff3.gz", destfile = "gencode.v25.annotation.gff3.gz")
txdb <- makeTxDbFromGFF("gencode.v25.annotation.gff3.gz")
txdb <- keepStandardChromosomes(txdb)
txdb
```

We will also need to convert the `gwasloc` object into a standard `GRanges` object:

```{r}
snps <- GRanges(snps)
```

Let's check if the `gwasloc` and `TxDb` object use the same notation for chromosomes:

```{r}
seqlevelsStyle(snps)
seqlevels(snps)
seqlevelsStyle(txdb)
seqlevels(txdb)
```

OK, they do.
Now we can annotate our SNPs to genes using the `VariantAnnotation` [@Obenchain2014] package:

```{r}
library(VariantAnnotation)
snps_anno <- locateVariants(snps, txdb, AllVariants())
snps_anno <- unique(snps_anno)
snps_anno
```

We lost all the metadata from the original `snps` object, but we can recover it using the `QUERYID` column in `snps_anno`.
We will only keep the SNP IDs and GWAS p-values:

```{r}
snps_metadata <- snps[snps_anno$QUERYID]
mcols(snps_anno) <- cbind(mcols(snps_metadata)[c("SNPS", "P-VALUE")], mcols(snps_anno))
snps_anno
```

We can visualise where these SNPs are located with `ggplot2` [@Wickham2009] (Figure \@ref(fig:barplot)):

```{r barplot, fig.cap = 'Barplot showing genomic locations associated with SLE variants.'}
library(ggplot2)
loc <- data.frame(table(snps_anno$LOCATION))
ggplot(data = loc, aes(x = reorder(Var1, -Freq), y = Freq)) +
  geom_bar(stat="identity")
```

As expected [@Maurano2012], the great majority of SNPs is located within introns and in intergenic regions.
For the moment, we will focus on SNPs that are either coding or in promoter and UTR regions, as these can be assigned to target genes rather unambiguously:

```{r}
snps_easy <- subset(snps_anno, LOCATION == "coding" | LOCATION == "promoter" | LOCATION == "threeUTR" | LOCATION == "fiveUTR")
snps_easy <- as.data.frame(snps_easy)
head(snps_easy)
```

Now we can check if any of the genes we found to be differentially expressed in SLE is also genetically associated with the disease:

```{r}
snps_easy_in_degs <- merge(degs, snps_easy, by.x = "gene_id", by.y = "GENEID", all = FALSE)
snps_easy_in_degs
```

So, we have `r length(unique(snps_easy_in_degs$gene_id))` genes showing differential expression in SLE that are also genetically associated with the disease.
While this is an interesting result, these hits are likely to be already well-known as potential SLE targets given their clear genetic association.

We will store essential information about these hits in a results `data.frame`:

```{r}
prioritised_hits <- unique(data.frame(
  snp_id = snps_easy_in_degs$SNPS,
  snp_pvalue = snps_easy_in_degs$P.VALUE,
  snp_location = snps_easy_in_degs$LOCATION,
  gene_id = snps_easy_in_degs$gene_id,
  gene_symbol = snps_easy_in_degs$symbol,
  gene_pvalue = snps_easy_in_degs$padj,
  gene_log2foldchange = snps_easy_in_degs$log2FoldChange))
prioritised_hits
```

## Use of regulatory genomic data to map intronic and intergenic SNPs to target genes

But what about all the SNPs in introns and intergenic regions?
Some of those might be regulatory SNPs affecting the expression level of their target gene through a distal enhancer.
Let's create a dataset of candidate regulatory SNPs that are either intronic or intergenic and remove the annotation obtained with `VariantAnnotation`:

```{r}
snps_hard <- subset(snps_anno, LOCATION == "intron" | LOCATION == "intergenic", select = c("SNPS", "P.VALUE", "LOCATION"))
snps_hard
```

### eQTL data

A well-established way to gain insights into target genes of regulatory SNPs is to use eQTL data, where correlations between genetic variants and expression of genes are computed across different tissues or cell types [@Albert2015].
We will use blood eQTL data from the GTEx consortium [@GTEx2017a].
To get the data, you will have to register and download the file `GTEx_Analysis_v7_eQTL.tar.gz` from the GTEx portal website [@GTEx2017b] to the current working directory:

```{r}
# uncomment the following line to extract the gzipped archive file
#untar("GTEx_Analysis_v7_eQTL.tar.gz")
gtex_blood <- read.delim(gzfile("GTEx_Analysis_v7_eQTL/Whole_Blood.v7.signif_variant_gene_pairs.txt.gz"), stringsAsFactors = FALSE)
head(gtex_blood)
```

We have to extract the SNPs locations from the ID used by GTEx:

```{r}
locs <- strsplit(gtex_blood$variant_id, "_")
gtex_blood$chr <- sapply(locs, "[", 1)
gtex_blood$start <- sapply(locs, "[", 2)
gtex_blood$end <- sapply(locs, "[", 2)
tail(gtex_blood)
```

We can then convert the `data.frame` into a `GRanges` object:

```{r}
gtex_blood <- makeGRangesFromDataFrame(gtex_blood, keep.extra.columns = TRUE)
gtex_blood
```

We also need to ensure that the style of the seqlevels is consistent with the previous objects:

```{r}
seqlevelsStyle(gtex_blood)
seqlevels(gtex_blood)
seqlevelsStyle(gtex_blood) <- "UCSC"
seqlevels(gtex_blood)
```

From the publication [@GTEx2017a], we know the genomic coordinates are mapped to genome reference GRCh37, so we will have to uplift them to GRCh38 using `rtracklayer` [@Lawrence2009] and a mapping ("chain") file.
The `R.utils` package [@Bengtsson2017] is required to extract the gzipped file: 

```{r}
library(rtracklayer)
library(R.utils)
# uncomment the following line to download file
#download.file("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/liftOver/hg19ToHg38.over.chain.gz", destfile = "hg19ToHg38.over.chain.gz")
# uncomment the following line to extract gzipped file
#gunzip("hg19ToHg38.over.chain.gz")
ch <- import.chain("hg19ToHg38.over.chain")
gtex_blood <- unlist(liftOver(gtex_blood, ch))
```

We will use the `GenomicRanges` package [@Lawrence2013] to compute the overlap between GWAS SNPs and eQTLs:

```{r}
library(GenomicRanges)
hits <- findOverlaps(snps_hard, gtex_blood)
snps_hard_in_gtex_blood = snps_hard[queryHits(hits)]
gtex_blood_with_snps_hard = gtex_blood[subjectHits(hits)]
mcols(snps_hard_in_gtex_blood) <- cbind(mcols(snps_hard_in_gtex_blood), mcols(gtex_blood_with_snps_hard))
snps_hard_in_gtex_blood <- as.data.frame(snps_hard_in_gtex_blood)
head(snps_hard_in_gtex_blood)
```

So, we have `r length(unique(snps_hard_in_gtex_blood$SNPS))` blood eQTL variants that are associated with SLE.
We can now check whether any of the genes differentially expressed in SLE is an *eGene*, a gene whose expression is influenced by an eQTL:

We note that gene IDs in GTEx are mapped to GENCODE v19 [@GTEx2017a], while we are using the newer v25 for the DEGs.
To match the gene IDs in the two objects, we will simply strip the last bit containing the GENCODE gene version, which effectively gives us Ensembl gene IDs:

```{r}
snps_hard_in_gtex_blood$ensembl_id <- sub("(ENSG[0-9]+)\\.[0-9]+", "\\1", snps_hard_in_gtex_blood$gene_id)
degs$ensembl_id <- sub("(ENSG[0-9]+)\\.[0-9]+", "\\1", degs$gene_id)
snps_hard_in_gtex_blood_in_degs <- merge(snps_hard_in_gtex_blood, degs, by = "ensembl_id", all = FALSE)
snps_hard_in_gtex_blood_in_degs
```

We can add these `r length(unique(snps_hard_in_gtex_blood_in_degs$gene_id.y))` genes to our list:

```{r}
prioritised_hits <- unique(rbind(prioritised_hits, data.frame(
  snp_id = snps_hard_in_gtex_blood_in_degs$SNPS,
  snp_pvalue = snps_hard_in_gtex_blood_in_degs$P.VALUE,
  snp_location = snps_hard_in_gtex_blood_in_degs$LOCATION,
  gene_id = snps_hard_in_gtex_blood_in_degs$gene_id.y,
  gene_symbol = snps_hard_in_gtex_blood_in_degs$symbol,
  gene_pvalue = snps_hard_in_gtex_blood_in_degs$padj,
  gene_log2foldchange = snps_hard_in_gtex_blood_in_degs$log2FoldChange)))
prioritised_hits
```

### FANTOM5 data

The FANTOM consortium profiled gene expression across a large panel of tissues and cell types using CAGE [@Fantom2014; @Andersson2014].
This technology allows mapping of transcription start sites (TSSs) and enhancer RNAs (eRNAs) genome-wide.
Correlations between these promoter and enhancer elements across a large panel of tissues and cell types can then be used to identify significant promoter - enhancer pairs, that we can use to map distal regulatory SNPs to target genes.

We can have a look at the enhancer - promoter correlation data in this way:

```{r}
# uncomment the following line to download the file
#download.file("http://enhancer.binf.ku.dk/presets/enhancer_tss_associations.bed", destfile = "enhancer_tss_associations.bed")
fantom <- read.delim("enhancer_tss_associations.bed", skip = 1, stringsAsFactors = FALSE)
head(fantom)
```

Everything we need is in the fourth column, `name`: genomic location of the enhancer, gene identifiers, correlation and significance.
We will use the `splitstackshape` package [@Mahto2014] to parse it:

```{r}
library(splitstackshape)
fantom <- as.data.frame(cSplit(fantom, splitCols = "name", sep = ";", direction = "wide"))
head(fantom)
```

Now we need to parse the `name_1` column which contains the genomic location of the enhancers:

```{r}
locs <- strsplit(as.character(fantom$name_1), "[:-]")
fantom$chr <- sapply(locs, "[", 1)
fantom$start <- as.numeric(sapply(locs, "[", 2))
fantom$end <- as.numeric(sapply(locs, "[", 3))
fantom$symbol <- fantom$name_3
fantom$corr <- sub("R:", "", fantom$name_4)
fantom$fdr <- sub("FDR:", "", fantom$name_5)
head(fantom)
```

We can select only those enhancer - promoter pairs with a decent level of correlation and significance and tidy the data at the same time:

```{r}
fantom <- unique(subset(fantom, subset = corr >= 0.25 & fdr < 1e-5, select = c("chr", "start", "end", "symbol")))
head(fantom)
```

Now we would like to check whether any of our candidate regulatory SNPs are falling in any of these enhancers.
To do this, we have to convert the `data.frame` into a `GRanges` object:

```{r}
fantom <- makeGRangesFromDataFrame(fantom, keep.extra.columns = TRUE)
fantom
```

Similar to the GTEx data, the FANTOM5 data is also mapped to GRCh37 [@Fantom2014], so we will have to uplift the GRCh37 coordinates to GRCh38:

```{r}
fantom <- unlist(liftOver(fantom, ch))
fantom
```

We can now compute the overlap between SNPs and enhancers:

```{r}
hits <- findOverlaps(snps_hard, fantom)
snps_hard_in_fantom = snps_hard[queryHits(hits)]
fantom_with_snps_hard = fantom[subjectHits(hits)]
mcols(snps_hard_in_fantom) <- cbind(mcols(snps_hard_in_fantom), mcols(fantom_with_snps_hard))
snps_hard_in_fantom <- as.data.frame(snps_hard_in_fantom)
snps_hard_in_fantom
```

We note that some of the SNPs are assigned to more than one gene.
This is because enhancers are promiscuous and can regulate multiple genes.

We can now check if any of these genes is differentially expressed in our RNA-seq data:

```{r}
snps_hard_in_fantom_in_degs <- merge(snps_hard_in_fantom, degs, by = "symbol", all = FALSE)
snps_hard_in_fantom_in_degs
```

We have identified `r length(unique(snps_hard_in_fantom_in_degs$gene_id))` genes whose putative enhancers contain SLE GWAS SNPs.
Let's add these to our list:

```{r}
prioritised_hits <- unique(rbind(prioritised_hits, data.frame(
  snp_id = snps_hard_in_fantom_in_degs$SNPS,
  snp_pvalue = snps_hard_in_fantom_in_degs$P.VALUE,
  snp_location = snps_hard_in_fantom_in_degs$LOCATION,
  gene_id = snps_hard_in_fantom_in_degs$gene_id,
  gene_symbol = snps_hard_in_fantom_in_degs$symbol,
  gene_pvalue = snps_hard_in_fantom_in_degs$padj,
  gene_log2foldchange = snps_hard_in_fantom_in_degs$log2FoldChange)))
prioritised_hits
```

### Promoter Capture Hi-C data

More recently, chromatin interaction data was generated across 17 human primary blood cell types [@Javierre2016].
More than 30,000 promoter baits were used to capture promoter-interacting regions genome-wide.
These regions were then mapped to enhancer based on the Ensembl Regulatory Build [@Zerbino2015] and can be accessed in the supplementary data of the paper:

```{r}
# uncomment the following line to download file
#download.file("http://www.cell.com/cms/attachment/2086554122/2074217047/mmc4.zip", destfile = "mmc4.zip")
# uncomment the following lines to extract zipped files
#unzip("mmc4.zip")
#unzip("DATA_S1.zip")
pchic <- read.delim("ActivePromoterEnhancerLinks.tsv", stringsAsFactors = FALSE)
head(pchic)
```

In this case, we will have to map the promoter baits to genes first.
We can do this by converting the baits to a `GRanges` object and then using the `TxDb` object we previously built to extract positions of transcription start sites (TSSs):

```{r}
baits <- GRanges(seqnames = pchic$baitChr, ranges = IRanges(start = pchic$baitSt, end = pchic$baitEnd))
tsss <- promoters(txdb, upstream = 0, downstream = 1, columns = "gene_id")
hits <- nearest(baits, tsss)
baits$gene_id <- unlist(tsss[hits]$gene_id)
baits
```

Now we can create a `GRanges` object of the enhancers in the promoter capture Hi-C data with the bait annotation attached:

```{r}
pchic <- GRanges(seqnames = pchic$oeChr, ranges = IRanges(start = pchic$oeSt, end = pchic$oeEnd), gene_id = baits$gene_id)
pchic <- unique(pchic)
pchic
```

Next, we basically repeat the steps we have taken when working with the FANTOM5 data to find SLE GWAS SNPs overlapping with these enhancers:

```{r}
hits <- findOverlaps(snps_hard, pchic)
snps_hard_in_pchic = snps_hard[queryHits(hits)]
pchic_with_snps_hard = pchic[subjectHits(hits)]
mcols(snps_hard_in_pchic) <- cbind(mcols(snps_hard_in_pchic), mcols(pchic_with_snps_hard))
snps_hard_in_pchic <- as.data.frame(snps_hard_in_pchic)
snps_hard_in_pchic
```

We check if any of these enhancer containing SLE variants are known to putatively regulate genes differentially expressed in SLE:

```{r}
snps_hard_in_pchic_in_degs <- merge(snps_hard_in_pchic, degs, by = "gene_id", all = FALSE)
snps_hard_in_pchic_in_degs
```

And finally we add these `r length(unique(snps_hard_in_pchic_in_degs$gene_id))` genes to our list.
These are our final results:

```{r}
prioritised_hits <- unique(rbind(prioritised_hits, data.frame(
  snp_id = snps_hard_in_pchic_in_degs$SNPS,
  snp_pvalue = snps_hard_in_pchic_in_degs$P.VALUE,
  snp_location = snps_hard_in_pchic_in_degs$LOCATION,
  gene_id = snps_hard_in_pchic_in_degs$gene_id,
  gene_symbol = snps_hard_in_pchic_in_degs$symbol,
  gene_pvalue = snps_hard_in_pchic_in_degs$padj,
  gene_log2foldchange = snps_hard_in_pchic_in_degs$log2FoldChange)))
prioritised_hits
```

# Conclusions

In this Bioconductor workflow we have used several packages and datasets downloaded from genomic projects and publications to show how regulatory genomic data can be used to annotate significant hits from GWASs and provide an intermediate layer connecting genetic and transcriptomic data.
Overall, we identified `r length(unique(prioritised_hits$snp_id))` SLE-associated SNPs that we mapped to `r length(unique(prioritised_hits$gene_id))` genes differentially expressed in SLE, using eQTL data [@GTEx2017a] and enhancer - promoter relationships from CAGE [@Fantom2014] and promoter capture Hi-C experiments [@Javierre2016].

While simplified, the workflow also demonstrates some real-world challenges encountered when working with genomic data from different sources, such as different genome references and gene annotation conventions as well as parsing of files with custom formats into Bioconductor-compatible objects.
As the sample size and power of GWASs and gene expression studies continue to increase, it will become more and more challenging to identify truly significant hits and interpret them.
The use of regulatory genomics data as presented here can be an important skill and tool to gain insights into large biomedical datasets and help in the identification of biomarkers and therapeutic targets. 

# Abbreviations

CAGE: cap analysis of gene expression  
eQTL: expression quantitative trait locus  
GWAS: genome-wide association study  
PheWAS: phenome-wide association study  
SLE: systemic lupus erythematosus  
SNP: single nucleotide polymorphism  

# Data and software availability

Download links for all datasets are part of the workflow.
Software packages required to reproduce the analysis can be installed as part of the workflow.
Code is available at https://github.com/enricoferrero/bioconductor-regulatory-genomics-workflow.

# Competing interests

EF is a full time employee of GSK.

# Grant information

The author declared that no grants were involved in supporting this work.

# References
